src/main/java/org/opensearch/security/dlic/rest/api/ConfigUpgradeApiAction.java

```diff
@@ -0,0 +1,395 @@
+                    @Override
+                    public Object[] params() {
+                        return params;
+                    }
+
+                    @Override
+                    public Settings settings() {
+                        return securityApiDependencies.settings();
+                    }
+
+                    @Override
+                    public Map<String, DataType> allowedKeys() {
+                        return Map.of(REQUEST_PARAM_CONFIGS_KEY, DataType.ARRAY);
+                    }
+                });
+            }
+        };
+    }
+
+    /** More permissions validation that default ContentValidator  */
+    static class ConfigUpgradeContentValidator extends RequestContentValidator {
+
+        protected ConfigUpgradeContentValidator(final ValidationContext validationContext) {
+            super(validationContext);
+        }
+
+        @Override
+        public ValidationResult<JsonNode> validate(final RestRequest request, final JsonNode jsonContent) throws IOException {
+            return validateContentSize(jsonContent);
+        }
+    }
+
+    /** Tranforms config changes from a raw PATCH into simplier view */
+    static class ConfigItemChanges {
+
+        private final CType config;
+        private final Map<String, List<String>> itemsGroupedByOperation;
+
+        public ConfigItemChanges(final CType config, final JsonNode differences) {
+            this.config = config;
+            this.itemsGroupedByOperation = classifyChanges(differences);
+        }
+
+        public boolean hasChanges() {
+            return !itemsGroupedByOperation.isEmpty();
+        }
+
+        public void addToNode(final ObjectNode node) {
+            final var allOperations = JsonNodeFactory.instance.objectNode();
+            itemsGroupedByOperation.forEach((operation, items) -> {
+                final var arrayNode = allOperations.putArray(operation);
+                items.forEach(arrayNode::add);
+            });
+            node.set(config.toLCString(), allOperations);
+        }
+
+        private static Map<String, List<String>> classifyChanges(final JsonNode differences) {
+            final var items = new HashMap<String, String>();
+            differences.forEach(node -> {
+                final var item = pathRoot(node);
+                final var operation = node.get("op").asText();
+                if (items.containsKey(item) && !items.get(item).equals(operation)) {
+                    items.put(item, "modify");
+                } else {
+                    items.put(item, operation);
+                }
+            });
+
+            final var itemsGroupedByOperation = items.entrySet()
+                .stream()
+                .collect(Collectors.groupingBy(Map.Entry::getValue, Collectors.mapping(Map.Entry::getKey, Collectors.toList())));
+            return itemsGroupedByOperation;
+        }
+    }
+}
```

Comment:
Could you add java-doc comments for all methods? I understand that the names are comprehensive but a brief description should make this class easy to follow.

Reply:
Could you add comments to specific methods that you think are unclear? I can add comments to them but I'd rather not generically fill them in.